File: 백준/Gold/7569. 토마토/토마토.py
# 7569 토마토
import sys
from collections import deque

M, N, H = map(int, sys.stdin.readline().split())

box = [[[] for _ in range(N)] for _ in range(H)]

for i in range(H):
    for j in range(N):
        temp = list(map(int, sys.stdin.readline().split()))
        for k in range(M):
            box[i][j].append(temp[k])

moving = [[1, 0, 0], [-1, 0, 0], [0, 0, -1], [0, 0, 1], [0, 1, 0], [0, -1, 0]]  # 위,아래,왼쪽,오른쪽,앞,뒤


def bfs():
    count = 0
    Q = deque()
    for height in range(H):
        for row in range(N):
            for col in range(M):
                if box[height][row][col] == 1:
                    Q.append((height, row, col))

    while Q:
        length = len(Q)
        flag = False
        for _ in range(length):
            height, row, col = Q.popleft()
            for move in moving:
                next_height = height + move[0]
                next_row = row + move[1]
                next_col = col + move[2]
                if (H > next_height >= 0 and N > next_row >= 0 and M > next_col >= 0
                        and box[next_height][next_row][next_col] == 0):
                    box[next_height][next_row][next_col] = 1
                    Q.append((next_height, next_row, next_col))
                    flag = True
        if flag:
            count += 1

    for height in range(H):
        for row in range(N):
            for col in range(M):
                if box[height][row][col] == 0:
                    print(-1)
                    return
    print(count)


bfs()

File: 백준/Gold/7576. 토마토/토마토.java
//  토마토

import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] s = br.readLine().split(" ");
        int M = Integer.parseInt(s[0]);
        int N = Integer.parseInt(s[1]);
        int[][] prob = new int[N][M];

        Queue<Point> Q = new LinkedList<>();

        for (int i = 0; i < N; i++) {
            String[] line = br.readLine().split(" ");
            for (int j = 0; j < M; j++) {
                prob[i][j] = Integer.parseInt(line[j]);
                if (prob[i][j] == 1)
                    Q.add(new Point(i, j));
            }
        }

        int[][] moving = {{1, 0}, {-1, 0}, {0, -1}, {0, 1}};
        int count = -1;
        while (!Q.isEmpty()) {

            count++;
            int size = Q.size();
            for (int i = 0; i < size; i++) {
                Point p = Q.poll();
                for (int[] move : moving) {
                    int nextRow = p.row + move[0];
                    int nextCol = p.col + move[1];
                    if (nextRow >= 0 && nextRow < N && nextCol >= 0 && nextCol < M && prob[nextRow][nextCol] == 0) {
                        prob[nextRow][nextCol] = 1;
                        Q.add(new Point(nextRow, nextCol));
                    }
                }
            }

        }

        outer:
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (prob[i][j] == 0) {
                    count = -1;
                    break outer;
                }
            }
        }

        System.out.println(count);

    }

    private static class Point {
        int row;
        int col;

        public Point(int x, int y) {
            this.row = x;
            this.col = y;
        }
    }
}
